packages/zoltan2/core/src/algorithms/color/Zoltan2_AlgHybrid2GL.hpp

```diff
@@ -0,0 +1,1532 @@
+      
+      //this view keeps track of the size of verts_to_send.
+      Kokkos::View<size_t*, device_type> verts_to_send_size("verts to send size",1);
+      Kokkos::View<size_t*, device_type, Kokkos::MemoryTraits<Kokkos::Atomic>> verts_to_send_size_atomic = verts_to_send_size;
+      typename Kokkos::View<size_t*, device_type>::HostMirror verts_to_send_size_host = create_mirror(verts_to_send_size);
+
+      verts_to_send_size_host(0) = 0;
+      Kokkos::deep_copy(verts_to_send_size, verts_to_send_size_host);
+
+      if(verbose) std::cout<<comm->getRank()<<": Constructing the boundary\n";
+      
+      //this function allocates and initializes the boundary_verts_dev view,
+      //sets the boundary_size accordingly, and adds vertices to the
+      //verts_to_send_atomic view, updating the size view as well.
+      constructBoundary(n_local, dist_offsets_dev, dist_adjs_dev, dist_offsets_host, dist_adjs_host, boundary_verts_dev, 
+	                boundary_size, verts_to_send_atomic, verts_to_send_size_atomic);
+      
+      //this boolean chooses which KokkosKernels algorithm to use.
+      //It was experimentally chosen for distance-1 coloring.
+      //Should be disregarded for distance-2 colorings.
+      bool use_vbbit = (global_max_degree < 6000);
+      //Done initializing, start coloring!
+      comm->barrier();
+      interior_time = timer();
+      total_time = timer();
+      //give the entire local graph to KokkosKernels to color
+      this->colorInterior(n_local, adjs_dev, offsets_dev, femv,adjs_dev,0,use_vbbit);
+      interior_time = timer() - interior_time;
+      comp_time = interior_time;
+      //get the color view from the FEMultiVector
+      Kokkos::View<int**, Kokkos::LayoutLeft> femvColors = femv->template getLocalView<memory_space>();
+      Kokkos::View<int*, device_type> femv_colors = subview(femvColors, Kokkos::ALL, 0);
+
+      //ghost_colors holds the colors of only ghost vertices.
+      //ghost_colors(0) holds the color of a vertex with LID n_local.
+      //To index this with LIDs, subtract n_local. If an LID is < n_local,
+      //it will not have a color stored in this view.
+      Kokkos::View<int*,device_type> ghost_colors("ghost color backups", n_ghosts);
+      comm->barrier();
+
+      //communicate the initial coloring.
+      if(verbose) std::cout<<comm->getRank()<<": communicating\n";
+      
+      //update the host views for the verts to send,
+      //doOwnedToGhosts needs host memory views, but doesn't
+      //change the values in them, so no need to copy afterwards
+      Kokkos::deep_copy(verts_to_send_host, verts_to_send_view);
+      Kokkos::deep_copy(verts_to_send_size_host, verts_to_send_size);
+      gno_t recv,sent;
+      comm_time = doOwnedToGhosts(mapOwnedPlusGhosts,n_local,verts_to_send_host,verts_to_send_size_host,femv,procs_to_send,sent,recv);
+      sentPerRound[0] = sent;
+      recvPerRound[0] = recv; 
+
+      Kokkos::parallel_for(n_ghosts, KOKKOS_LAMBDA(const int& i){
```

Comment:
Add comment:  what's happening here
Storing ghost colors so that can restore them after recoloring